File: Shared/Reading/WikiLinks.swift
import Foundation

/// A page that links to the current page with `[[…]]`.
nonisolated struct Backlink: Identifiable, Sendable, Equatable {
    let url: URL
    let title: String
    /// The line the link is on.
    let snippet: String

    var id: URL { url }
}

nonisolated enum WikiLinks {
    /// Reading mode turns `[[Page]]` into a link with this scheme, which Blad handles itself.
    static let urlScheme = "blad-page"

    /// `[[Page]]` or `[[Page|shown text]]`: group 1 is the page, group 2 the optional text.
    static let pattern = try! NSRegularExpression(pattern: #"\[\[([^\[\]\n|]+)(?:\|([^\[\]\n]+))?\]\]"#)

    /// Rewrites `[[Page]]` as a markdown link, so reading mode can render and open it.
    static func markdownLinks(in text: String) -> String {
        guard text.contains("[[") else { return text }
        let source = text as NSString
        var result = ""
        var last = 0
        for match in pattern.matches(in: text, range: NSRange(location: 0, length: source.length)) {
            result += source.substring(with: NSRange(location: last, length: match.range.location - last))
            let page = source.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespaces)
            let shown = match.range(at: 2).location == NSNotFound ? page : source.substring(with: match.range(at: 2))
            let encoded = page.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? page
            result += "[\(shown)](\(urlScheme):\(encoded))"
            last = NSMaxRange(match.range)
        }
        return result + source.substring(from: last)
    }

    static func backlinks(to title: String, excluding url: URL, in entries: [SearchEntry]) -> [Backlink] {
        entries.compactMap { entry in
            guard entry.url != url else { return nil }
            let text = entry.text as NSString
            for match in pattern.matches(in: entry.text, range: NSRange(location: 0, length: text.length)) {
                let target = (text.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespaces) as NSString).lastPathComponent
                guard target.compare(title, options: [.caseInsensitive, .diacriticInsensitive]) == .orderedSame else { continue }
                let line = text.substring(with: text.lineRange(for: match.range)).trimmingCharacters(in: .whitespacesAndNewlines)
                return Backlink(url: entry.url, title: entry.title, snippet: line)
            }
            return nil
        }
    }
}

File: windows/Blad.Core/Markdown/WikiLinks.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Blad.Core.Search;

namespace Blad.Core.Markdown;

/// <summary>A page that links to the current page with [[…]], and the line the link is on.</summary>
public sealed record Backlink(string Path, string Title, string Snippet);

public static partial class WikiLinks
{
    /// <summary>Reading mode turns [[Page]] into a link with this scheme, which Blad handles itself.</summary>
    public const string UrlScheme = "blad-page";

    /// <summary>[[Page]] or [[Page|shown text]]: group 1 is the page, group 2 the optional text.</summary>
    [GeneratedRegex(@"\[\[([^\[\]\n|]+)(?:\|([^\[\]\n]+))?\]\]")]
    public static partial Regex Pattern();

    public static string LinkTo(string page) => $"{UrlScheme}:{Uri.EscapeDataString(page)}";

    /// <summary>The page a <c>blad-page:</c> link points to, or null for any other link.</summary>
    public static string? PageName(string link) =>
        link.StartsWith(UrlScheme + ":", StringComparison.Ordinal)
            ? Uri.UnescapeDataString(link[(UrlScheme.Length + 1)..])
            : null;

    /// <summary>Page names match ignoring case and accents, so [[ideeen]] finds "Ideeën".</summary>
    public static bool SameName(string a, string b) =>
        string.Compare(a, b, CultureInfo.InvariantCulture, CompareOptions.IgnoreCase | CompareOptions.IgnoreNonSpace) == 0;

    public static IReadOnlyList<Backlink> Backlinks(string title, string excludingPath, IEnumerable<SearchEntry> entries)
    {
        var backlinks = new List<Backlink>();
        foreach (var entry in entries)
        {
            if (string.Equals(entry.Path, excludingPath, StringComparison.OrdinalIgnoreCase)) continue;
            foreach (Match match in Pattern().Matches(entry.Text))
            {
                var target = Path.GetFileName(match.Groups[1].Value.Trim());
                if (!SameName(target, title)) continue;
                backlinks.Add(new Backlink(entry.Path, entry.Title, LineAround(entry.Text, match.Index)));
                break;
            }
        }
        return backlinks;
    }

    private static string LineAround(string text, int index)
    {
        var start = text.LastIndexOf('\n', Math.Max(0, index - 1)) + 1;
        var end = text.IndexOf('\n', index);
        return text[start..(end < 0 ? text.Length : end)].Trim();
    }
}
